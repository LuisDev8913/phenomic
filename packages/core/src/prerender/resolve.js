import createQuery from "@phenomic/api-client/lib/query";

const debug = require("debug")("phenomic:core:prerender:resolve");

const resolveURLsForDynamicParams = async function(
  phenomicFetch: PhenomicFetch,
  route: PhenomicRoute
) {
  const collectionConfig = typeof route.collection === "string"
    ? { collection: route.collection }
    : route.collection;
  debug("route", route.path);
  if (!collectionConfig) {
    debug("no valid collection", route.collection);
    return route;
  }
  // If the path doesn't contain any kind of parameter, no need to
  // iterate over the collection
  if (!route.path.includes("*") && !route.path.includes(":")) {
    debug("not a dynamic route");
    return route;
  }
  debug(
    `fetching collection '${collectionConfig.collection
      ? collectionConfig.collection
      : Object.keys(collectionConfig).join(",")}' for route '${route.path}'`
  );
  // @todo memoize for perfs and avoid uncessary call
  const collection = await phenomicFetch(createQuery(collectionConfig));
  debug(`collection fetched. ${collection.list.length} items`);
  const path = route.path || "*";
  return collection.list.map(item => {
    return {
      ...route,
      path: path.replace(/\*/, item.key),
      params: { splat: item.key }
    };
  });
};

const findPaginatedQuery = function(queries) {
  const key = Object.keys(queries).find(key =>
    queries[key].hasOwnProperty("after")
  );
  if (!key) {
    return null;
  }
  return queries[key];
};

const resolveNextURLsInPagination = async function(
  path: string,
  query: PhenomicQueryConfig,
  phenomicFetch: PhenomicFetch,
  urls: Array<string> = []
) {
  urls = [
    ...urls,
    path.replace("/:after", query.after ? "/" + query.after : "")
  ];
  debug(path, query.after);
  const nextPage = await phenomicFetch(createQuery(query));
  if (nextPage.hasNextPage) {
    debug(path, "have a next page");
    return resolveNextURLsInPagination(
      path,
      { ...query, after: nextPage.next },
      phenomicFetch,
      urls
    );
  } else {
    return urls;
  }
};

const resolvePaginatedURLs = async function(
  fetch: PhenomicFetch,
  route: PhenomicRoute
) {
  if (!route.paginated) {
    debug(route.path, "is not paginated");
    return route.path;
  }
  if (!route.component.getQueries) {
    debug(route.path, "is paginated but have no queries");
    return route.path;
  }
  const initialRouteParams: Object = route.params || {};
  const initialRouteQuery = route.component.getQueries({
    params: initialRouteParams
  });
  const query = findPaginatedQuery(initialRouteQuery);
  if (!query) {
    debug(route.path, "is paginated with queries, but no :after param found");
    return route.path;
  }
  return resolveNextURLsInPagination(route.path, query, fetch);
};

const flatten = (array: Array<any>) => {
  const flattenedArray = [];
  array.forEach(item => {
    Array.isArray(item)
      ? flattenedArray.push(...flatten(item))
      : flattenedArray.push(item);
  });

  return flattenedArray;
};

const normalizePath = (path: string) => path.replace(/^\//, "");

const resolveURLsToPrerender = async function(
  routes: Array<PhenomicRoute>,
  fetch: PhenomicFetch
) {
  const dynamicRoutes = await Promise.all(
    routes.map(route => resolveURLsForDynamicParams(fetch, route))
  );
  const flattenedDynamicRoutes = flatten(dynamicRoutes);
  const filtredDynamicRoutes = flattenedDynamicRoutes.filter(url => {
    if (url.path && url.path.includes("*") && !url.collection) {
      debug(
        `${url.path} is including a '*' but it has not been resolved: url is skipped`
      );
      return false;
    }
    return true;
  });
  const paginatedURLs = await Promise.all(
    filtredDynamicRoutes.map(route => resolvePaginatedURLs(fetch, route))
  );
  const normalizedURLs = flatten(paginatedURLs).map(normalizePath);
  const uniqsNormalizedPath = [...new Set(normalizedURLs)];
  return uniqsNormalizedPath;
};

export default resolveURLsToPrerender;
