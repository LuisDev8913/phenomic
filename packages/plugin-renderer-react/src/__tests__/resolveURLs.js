/* eslint-disable import/first */
jest.mock("@phenomic/api-client/lib/fetch", () => () =>
  Promise.resolve({
    list: []
  })
);

import resolve from "../resolveURLs.js";

const NoopComponent = {
  fakeComponent: true // to pass test on component being valid
};

it("should be able to generate a single url", async () => {
  expect(
    await resolve({
      routes: [
        {
          path: "/",
          component: NoopComponent
        }
      ]
    })
  ).toMatchSnapshot();
});

it("should be able to generate a multiple static url", async () => {
  expect(
    await resolve({
      routes: [
        {
          path: "/",
          component: NoopComponent
        },
        {
          path: "/test",
          component: NoopComponent
        }
      ]
    })
  ).toMatchSnapshot();
});
