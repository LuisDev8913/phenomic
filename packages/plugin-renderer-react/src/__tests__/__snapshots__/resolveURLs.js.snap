// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be able to generate a multiple static url 1`] = `
Array [
  "",
  "test",
]
`;

exports[`should be able to generate a single url 1`] = `
Array [
  "",
]
`;
